lib/index.js

```diff
@@ -679,33 +679,45 @@ function promiseStream(StreamCtor, promise) {
 
 function iteratorStream(StreamCtor, it) {
     return new StreamCtor(function (push, next) {
-        var iterElem, iterErr;
-        try {
-            iterElem = it.next();
-        }
-        catch (err) {
-            iterErr = err;
+        function pushIt(iterElem) {
+            if (iterElem.done) {
+                if (!_.isUndefined(iterElem.value)) {
+                    // generators can return a final
+                    // value on completion using return
+                    // keyword otherwise value will be
+                    // undefined
+                    push(null, iterElem.value);
+                }
+                push(null, _.nil);
+            }
+            else {
+                push(null, iterElem.value);
+                next();
+            }
         }
 
-        if (iterErr) {
-            push(iterErr);
-            push(null, _.nil);
-        }
-        else if (iterElem.done) {
-            if (!_.isUndefined(iterElem.value)) {
-                // generators can return a final
-                // value on completion using return
-                // keyword otherwise value will be
-                // undefined
-                push(null, iterElem.value);
+        try {
+            var iterElem = it.next();
+
+            if (_.isFunction(iterElem.then)) {
+                iterElem
+                    .then(pushIt)
+                    .catch(function(err) {
+                        // this is needed, but not sure why
+                        _.setImmediate(function() {
```

Comment:
Is this still necessary? I tried pulling your changes and ran the test without setImmediate, and they all passed.

Reply:
Yes, sorry that comment is poorly phrased:  I have "real" code that misbehaves when this is not present, but the tests are fine; I hope to add a test that demonstrates the problem, I just have to come up with one.

Reply:
A red herring I think, it was because of tomfoolery in some other part of code (I was using observe().forEach() without handling errors and that cause some.. oddness).

I've removed this.